Boardtest draft (#1)
aktole/oneAPI-samples: DirectProgramming/DPC++FPGA/PlatformTests/board_test/README.md

# Board Test
This design contains tests to check FPGA board interfaces and reports the following metrics:
1. Host to device global memory interface bandwidth
2. Kernel clock frequency
3. Kernel launch latency
4. Kernel to device global memory bandwidth

***Documentation***:  The [DPC++ FPGA Code Samples Guide](https://software.intel.com/content/www/us/en/develop/articles/explore-dpcpp-through-intel-fpga-code-samples.html) helps you to navigate the samples and build your knowledge of DPC++ for FPGA. <br>
The [oneAPI DPC++ FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide) is the reference manual for targeting FPGAs through DPC++. <br>
The [oneAPI Programming Guide](https://software.intel.com/en-us/oneapi-programming-guide) is a general resource for target-independent DPC++ programming.

| Optimized for                     | Description
---                                 |---

> | OS                                | Linux* Ubuntu* 18.04/20.04, RHEL*/CentOS* 8, SUSE* 15

Reviewer: why no windows support? what if someone wants to develop a windows platform?

Author: Adding Windows support

| Hardware                          | Intel® Programmable Acceleration Card (PAC) with Intel Arria® 10 GX FPGA <br> Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX) <br> Intel® FPGA 3rd party / custom platforms with oneAPI support <br> *__Note__: Intel® FPGA PAC hardware is only compatible with Ubuntu 18.04* 
| Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit
| What you will learn               | How to test board interfaces to ensure designed platform provides expected performance
| Time to complete                  | 30 minutes (not including compile time)

## Purpose
This test can be used to check FPGA board interfaces and measure host to device as well as kernel to global memory interface metrics. Custom platform developers can use this test as a starting point to validate custom platform interfaces. 

### Introduction to oneAPI FPGA Platform 

> Block diagram below shows an overview of oneAPI FPGA platform interfaces. The numbered arrows depict the following:

Reviewer: ```suggestion
The block diagram below shows an overview of how a oneAPI accelerator is implemented in an FPGA. The numbered arrows depict the following:
```

Instead of referring to numbered arrows can you draw the interfaces in the graphic (e.g. an arrow between host and device global memory that is superimposed on the path throuh PCIe IP and DMA)?

Author: Updated diagram to show arrows superimposed on the paths to be represented

* Path 1 + 2 depicts host to device global memory interface

> * Path 1 + 3 depicts host to kernel interface 

Reviewer: This is definitely confusing because there is literally a block in this figure that is labeled 'Kernel interface' :)

Author: Block name changed to kernel interface IP

* Path 4 shows kernel to device global memory interface

> oneAPI platform design consists of components operating at different clock domains. PCIe and external memories operate at a fixed frequency. Corresponding RTL IPs are parametrized to operate at these fixed frequencies by platform vendors. 

Reviewer: ```suggestion
A oneAPI accelerator consists of components operating at different clock domains. PCIe and external memories operate at a fixed frequency. Corresponding RTL IPs are parametrized to operate at these fixed frequencies by platform vendors. 
```

Can we use a different word than 'platform design' if we don't want to guide users to 'Platform Designer'? :)

Reviewer 2: We probably  should avoid creating new directories if we can. It is a pain. I would put this under: 
https://github.com/oneapi-src/oneAPI-samples/tree/master/DirectProgramming/DPC%2B%2BFPGA/Tutorials/Tools
or
https://github.com/oneapi-src/oneAPI-samples/tree/master/DirectProgramming/DPC%2B%2BFPGA/Tutorials/Features

Reviewer 3: I'll side with Paul and say that this should go under the "Reference Designs" folder rather than creating a separate top folder

Author: Moved to Reference Designs

Kernel clock frequency is variable and is calculated as a part of oneAPI offline compilation flow for FPGAs.  `board_test` measures the frequency that the kernel is running at in the FPGA and compares this to compiled kernel clock frequency. [Test Details](#test-details) section provides more information.

> *Note:* The block diagram shown below is an overview of a typical oneAPI FPGA platform. Please refer to oneAPI platform vendor documentation for more details on platform components. 

Reviewer: which modules in this graphic are 'platform components'?

Also, can you link an example of platform vendor documentation for the D5005 board that discusses platform components?

Author: Added a brief description of a oneAPI BSP in the introduction section to help understand the block diagram better

> ![](oneapi_fpga_platform.PNG)

Reviewer: Can the PCIe IP and DMA (and part of Kernel Interface) be green instead of purple? Purple and blue are kind of hard to distinguish, and differentiating between green and blue aligns with how OFS categorizes 'blue bitstreams' and 'green bitstreams' in terms of what the end user can modify. I realize that everything in this graphic is technically 'blue' as far as OFS is concerned.

Author: Diagram color scheme updated

### Test Details
The complete test exercises following interfaces in a platform:

> 1. **Host to device global memory interface:** This interface is checked by performing explicit data movement between the host and device global memory. Host to device global memory bandwidth is measured and reported. As a part of this interface check, unaligned data transfers are also performed to verify non-DMA transfers complete successfully.

Reviewer: ```
unaligned data transfers are also performed to verify non-DMA transfers complete successfully
```

non-dma or non-aligned dma?

Reviewer 2: I think, based on my understanding of our MMD, that unaligned buffers will still be DMAed. I think we will copy them into a temp aligned AND pinned buffer and then DMA. But I think this is still a decent test to have.

> 2. **Kernel to device global memory interface:** This interfaces is checked by performing kernel to memory data transfers using simple read & write kernels. Kernel to memory bandwidth is measured and reported.<br> 

Reviewer: ```suggestion
2. **Kernel to device global memory interface:** This interface is checked by performing kernel to memory data transfers using simple read and write kernels. Kernel to memory bandwidth is measured and reported.<br> 
```

Author: done

*Note:* This test currently does not support SYCL Unified Shared Memory (USM). For testing USM interface, please use [Simple host streaming sample](https://github.com/oneapi-src/oneAPI-samples/tree/master/DirectProgramming/DPC%2B%2BFPGA/Tutorials/DesignPatterns/simple_host_streaming).

> 3. **Host to kernel interface:** The test ensures the host to kernel communication is correct and host can launch a kernel successfully. It also measures the roundtrip kernel launch latency and throughput (number of kernel/ms) of single task no-operation kernels.

Reviewer: ```suggestion
3. **Host to kernel interface:** The test ensures the host to kernel communication is correct and that the host can launch a kernel successfully. It also measures the roundtrip kernel launch latency and throughput (number of kernels/ms) of single task no-operation kernels.
```

Author: done

> In addition to above interface checks, the test also measures the following metric:

Reviewer: ```suggestion
In addition to the above interface checks, the test also measures the 
```

A numbered list doesn't make sense if there is only one thing in it :) I think the phrasing in that description should be changed too.

Author: done

> 1. **Kernel clock frequency:** The test measures the frequency that the programmed kernel is running at on the FPGA device and reports it. By default, this test fails if the measured frequency is not within 2% of the compiled frequency.<br>

Reviewer: ```suggestion
**Kernel clock frequency:** The test measures the frequency that the programmed kernel is running at on the FPGA device and reports it. By default, this test fails if the measured frequency is not within 2% of the compiled frequency.<br>
```

Author: done

*Note:* The test allows overriding this failure, however, this is may lead to functional errors and is not recommended. The override option is provided to allow debug in case where platform design changes are done to force kernel to run at slower clock (this is not a common use-case). Usage section describes how to override this option if required. 

> Above interface checks are divided into six sub-tests. By default, all tests are run. You can choose to run a single test using `-test=<test number>` option. Please see [Running the Sample](#running-the-sample) for test usage instructions. 

Reviewer: ```suggestion
The complete test is divided into six sub-tests. By default, all tests are run. You can choose to run a single test using `-test=<test number>` option. Please see [Running the Sample](#running-the-sample) for test usage instructions. 
```

The list below should clarify which test exercises which interfaces.

Author: done

When compiling for FPGA hardware, it is recommended to increase the job timeout to 12h.

### On a Linux* System

1. Generate the `Makefile` by running `cmake`.
     ```
   mkdir build
   cd build
   ```
   To compile for the Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX), run `cmake` using the command:
    ```
    cmake ..
   ```
   Alternatively, to compile for the Intel® PAC with Intel Arria® 10 GX FPGA, run `cmake` using the command:

   ```
   cmake .. -DFPGA_BOARD=intel_a10gx_pac:pac_a10
   ```
   You can also compile for a custom FPGA platform. Ensure that the board support package is installed on your system. Then run `cmake` using the command:
   ```
   cmake .. -DFPGA_BOARD=<board-support-package>:<board-variant>
   ```

2. Compile the design through the generated `Makefile`. The following build targets are provided, matching the recommended development flow:

   * Compile for emulation (compiles quickly, targets emulated FPGA device):
      ```

>       make fpga_emu

Reviewer: I tried this in ARC using 2021.3 and 2021.4, and I got this mass of warnings:

```
board_test $> mkdir build

board_test $> cd build/
build $> cmake ..
-- The CXX compiler identification is Clang 13.0.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /p/psg/swip/hld/r/sycl/rel/20210619/linux64/bin/dpcpp - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- FPGA_BOARD was not specified.                    
Configuring the design to run on the default FPGA board intel_s10sx_pac:pac_s10 (Intel(R) PAC with Intel Stratix(R) 10 SX FPGA).                     
Please refer to the README for information on board selection.
-- Configuring done
-- Generating done
-- Build files have been written to: /p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/build

build $> make fpga_emu
[ 50%] Building CXX object src/CMakeFiles/board_test.fpga_emu.dir/board_test.cpp.o
In file included from /p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.cpp:34:
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:22:35: warning: 'INTEL' is deprecated: use 'ext::intel' instead [-Wdeprecated-declarations]
using SendertoReceiverPipe = INTEL::pipe<  // Defined in the SYCL headers
                                  ^
/p/psg/swip/hld/r/sycl/rel/20210915/linux64/include/sycl/ext/intel/fpga_reg.hpp:49:11: note: 'INTEL' has been explicitly marked deprecated here
namespace __SYCL2020_DEPRECATED("use 'ext::intel' instead") INTEL {
          ^
/p/psg/swip/hld/r/sycl/rel/20210915/linux64/bin/../include/sycl/CL/sycl/detail/defines_elementary.hpp:52:40: note: expanded from macro '__SYCL2020_DEPRECATED'
#define __SYCL2020_DEPRECATED(message) __SYCL_DEPRECATED(message)
                                       ^
/p/psg/swip/hld/r/sycl/rel/20210915/linux64/bin/../include/sycl/CL/sycl/detail/defines_elementary.hpp:43:38: note: expanded from macro '__SYCL_DEPRECATED'
#define __SYCL_DEPRECATED(message) [[deprecated(message)]]
                                     ^
In file included from /p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.cpp:34:
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:735:5: warning: attribute 'cl::reqd_work_group_size' is deprecated [-Wdeprecated-attributes]
                [[cl::reqd_work_group_size(1, 1, kReqdWgSize)]]{});
                  ^
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:735:5: note: did you mean to use 'sycl::reqd_work_group_size' instead?
                [[cl::reqd_work_group_size(1, 1, kReqdWgSize)]]{});
                  ^~
                  sycl                                                                                                                                                                                                                                                       
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:1572:38: warning: attribute 'cl::reqd_work_group_size' is deprecated [-Wdeprecated-attributes]
                                     cl::reqd_work_group_size(1, 1, kWGSize)]] {
                                     ^
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:1572:38: note: did you mean to use 'sycl::reqd_work_group_size' instead?
                                     cl::reqd_work_group_size(1, 1, kWGSize)]] {
                                     ^~
                                     sycl                                                                                                                                                                                                                                    
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:1584:38: warning: attribute 'cl::reqd_work_group_size' is deprecated [-Wdeprecated-attributes]
                                     cl::reqd_work_group_size(1, 1, kWGSize)]] {
                                     ^
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:1584:38: note: did you mean to use 'sycl::reqd_work_group_size' instead?
                                     cl::reqd_work_group_size(1, 1, kWGSize)]] {
                                     ^~
                                     sycl                                                                                                                                                                                                                                    
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:1603:38: warning: attribute 'cl::reqd_work_group_size' is deprecated [-Wdeprecated-attributes]
                                     cl::reqd_work_group_size(1, 1, kWGSize)]] {
                                     ^
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:1603:38: note: did you mean to use 'sycl::reqd_work_group_size' instead?
                                     cl::reqd_work_group_size(1, 1, kWGSize)]] {
                                     ^~
                                     sycl                                                                                                                                                                                                                                    
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.cpp:83:8: warning: 'INTEL' is deprecated: use 'ext::intel' instead [-Wdeprecated-declarations]
  INTEL::fpga_emulator_selector device_selector;
       ^
/p/psg/swip/hld/r/sycl/rel/20210915/linux64/include/sycl/ext/intel/fpga_reg.hpp:49:11: note: 'INTEL' has been explicitly marked deprecated here
namespace __SYCL2020_DEPRECATED("use 'ext::intel' instead") INTEL {
          ^
/p/psg/swip/hld/r/sycl/rel/20210915/linux64/bin/../include/sycl/CL/sycl/detail/defines_elementary.hpp:52:40: note: expanded from macro '__SYCL2020_DEPRECATED'
#define __SYCL2020_DEPRECATED(message) __SYCL_DEPRECATED(message)
                                       ^
/p/psg/swip/hld/r/sycl/rel/20210915/linux64/bin/../include/sycl/CL/sycl/detail/defines_elementary.hpp:43:38: note: expanded from macro '__SYCL_DEPRECATED'
#define __SYCL_DEPRECATED(message) [[deprecated(message)]]
                                     ^
6 warnings generated.
In file included from /p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.cpp:34:
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:22:35: warning: 'INTEL' is deprecated: use 'ext::intel' instead [-Wdeprecated-declarations]
using SendertoReceiverPipe = INTEL::pipe<  // Defined in the SYCL headers
                                  ^
/p/psg/swip/hld/r/sycl/rel/20210915/linux64/include/sycl/ext/intel/fpga_reg.hpp:49:11: note: 'INTEL' has been explicitly marked deprecated here
namespace __SYCL2020_DEPRECATED("use 'ext::intel' instead") INTEL {
          ^
/p/psg/swip/hld/r/sycl/rel/20210915/linux64/bin/../include/sycl/CL/sycl/detail/defines_elementary.hpp:52:40: note: expanded from macro '__SYCL2020_DEPRECATED'
#define __SYCL2020_DEPRECATED(message) __SYCL_DEPRECATED(message)
                                       ^
/p/psg/swip/hld/r/sycl/rel/20210915/linux64/bin/../include/sycl/CL/sycl/detail/defines_elementary.hpp:43:38: note: expanded from macro '__SYCL_DEPRECATED'
#define __SYCL_DEPRECATED(message) [[deprecated(message)]]
                                     ^
In file included from /p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.cpp:34:
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:735:5: warning: attribute 'cl::reqd_work_group_size' is deprecated [-Wdeprecated-attributes]
                [[cl::reqd_work_group_size(1, 1, kReqdWgSize)]]{});
                  ^
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:735:5: note: did you mean to use 'sycl::reqd_work_group_size' instead?
                [[cl::reqd_work_group_size(1, 1, kReqdWgSize)]]{});
                  ^~
                  sycl                                                                                                                                                                                                                                                       
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:1572:38: warning: attribute 'cl::reqd_work_group_size' is deprecated [-Wdeprecated-attributes]
                                     cl::reqd_work_group_size(1, 1, kWGSize)]] {
                                     ^
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:1572:38: note: did you mean to use 'sycl::reqd_work_group_size' instead?
                                     cl::reqd_work_group_size(1, 1, kWGSize)]] {
                                     ^~
                                     sycl                                                                                                                                                                                                                                    
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:1584:38: warning: attribute 'cl::reqd_work_group_size' is deprecated [-Wdeprecated-attributes]
                                     cl::reqd_work_group_size(1, 1, kWGSize)]] {
                                     ^
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:1584:38: note: did you mean to use 'sycl::reqd_work_group_size' instead?
                                     cl::reqd_work_group_size(1, 1, kWGSize)]] {
                                     ^~
                                     sycl                                                                                                                                                                                                                                    
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:1603:38: warning: attribute 'cl::reqd_work_group_size' is deprecated [-Wdeprecated-attributes]
                                     cl::reqd_work_group_size(1, 1, kWGSize)]] {
                                     ^
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp:1603:38: note: did you mean to use 'sycl::reqd_work_group_size' instead?
                                     cl::reqd_work_group_size(1, 1, kWGSize)]] {
                                     ^~
                                     sycl                                                                                                                                                                                                                                    
/p/psg/swip/w/whitepau/oneAPI_Experiments/anuja-oneAPI-samples/DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.cpp:83:8: warning: 'INTEL' is deprecated: use 'ext::intel' instead [-Wdeprecated-declarations]
  INTEL::fpga_emulator_selector device_selector;
       ^
/p/psg/swip/hld/r/sycl/rel/20210915/linux64/include/sycl/ext/intel/fpga_reg.hpp:49:11: note: 'INTEL' has been explicitly marked deprecated here
namespace __SYCL2020_DEPRECATED("use 'ext::intel' instead") INTEL {
          ^
/p/psg/swip/hld/r/sycl/rel/20210915/linux64/bin/../include/sycl/CL/sycl/detail/defines_elementary.hpp:52:40: note: expanded from macro '__SYCL2020_DEPRECATED'
#define __SYCL2020_DEPRECATED(message) __SYCL_DEPRECATED(message)
                                       ^
/p/psg/swip/hld/r/sycl/rel/20210915/linux64/bin/../include/sycl/CL/sycl/detail/defines_elementary.hpp:43:38: note: expanded from macro '__SYCL_DEPRECATED'
#define __SYCL_DEPRECATED(message) [[deprecated(message)]]
                                     ^
6 warnings generated.
[100%] Linking CXX executable ../board_test.fpga_emu
Platform name: Intel(R) FPGA Emulation Platform for OpenCL(TM)
Device name: Intel(R) FPGA Emulation Device
Driver version: 2021.12.9.0.15_160000
OpenCL program was successfully created from SPIR-V file /tmp/arc_355788651/board_test-22cdac-be91ce.spv
Using build options: -g  -I "/tmp/arc_355788651"

build log:
Compilation started
Compilation done
Linking started
Linking done
Device build started
Options used by backend compiler: -g  -I "/tmp/arc_355788651"
Device build done
Kernel <_ZTS10NopNDRange> was successfully vectorized (8)
Kernel <_ZTS13NopSingleTask> was successfully vectorized (8)
Kernel <_ZTS12KernelSender> was not vectorized
Kernel <_ZTS14KernelReceiver> was not vectorized
Kernel <_ZTS18MemReadWriteStream> was successfully vectorized (8)
Kernel <_ZTS14MemWriteStream> was successfully vectorized (8)
Kernel <_ZTS13MemReadStream> was successfully vectorized (8)
Kernel <_ZTS25MemReadWriteStreamNDRange> was successfully vectorized (8)
Done.
OpenCL program binary file was successfully created: /tmp/arc_355788651/board_test-eb35b5-528b37.aocx
[100%] Built target board_test.fpga_emu
Scanning dependencies of target fpga_emu
[100%] Built target fpga_emu
build $> ./
CMakeFiles/          board_test.fpga_emu  src/                 
build $> 
```

Author: Cleaner compilers now, warnings resolved.
